requests-for-comments/2023-03-01-Temporal-RFC.md

```diff
@@ -0,0 +1,239 @@
+# RFC: Temporal at PostHog
+
+Author: James Greenhill
+
+Date: February 27, 2023 
+
+# For the reader
+
+Hi there, dear reader. I would love for you to scrutinize this RFC and the following RFC with more specifics on how we can build out our workflows. I’m looking for two things in particular from you though:
+
+- Is this a good idea?
+- Is getting this done feasible?
+
+I’m specifically looking for feedback from teams that are directly impacted by this, namely:
+
+- Infrastructure
+- Billing
+- Pipeline
+
+# Problem Statement
+
+We have a number of tasks that we’d like to run and have the confidence that they ran to completion. It would be even nicer to have some confidence that we were able to run it to completion exactly once despite hardware issues, deployments, or James going in and restarting a machine manually. So far we’ve built a few different ways to do this that are all quite different: Celery, Kafka, and Graphile. Each of these is a worker queue or worker queue adjacent that we have to maintain and keep flowing. The goal of this RFC is to propose a new, consolidated solution that we can use to focus more on our business logic (the product that we are actually building) and less on making sure that that logic actually runs. It’s a lot easier to reason about a product if you don’t have to think as much about it’s failure modes, retry logic, and timeouts.
+
+## Requirements
+
+We have a few products that are good candidates for a workflow framework. We’ll cover a few of the requirements here before we jump into any proposals just so that we can have a better understanding of what we are looking for.
+
+### Billing
+
+Billing is one of those things that is really tricky. Ideally you don’t want to bill someone twice (oof). You also don’t want to think you bill someone and then not bill them at all. Ideally you can kick off a flow and at the end of it have confidence that everything worked. The system was able to grab the usage metrics, apply the billing rate, and charge the user without having to double check the usage and the stripe api. Even though I still believe in trust, but verify. Any billing mistakes are generally pretty embarrassing.
+
+Requirements:
+
+- **Idempotent** or **exactly once** - no duplicates
+- **Retry-able** - If the thing fails, keep trying to get the money we are owed
+- **Resumable** - nice to pick up where you left off
+- **Composable** - Query ClickHouse. Calculate bill. Charge stripe.
+- **Parallelize-able** - We should be able to run multiple bills at the same time
+- **Timeouts** - We should be able to set timeouts on steps
+- **Cancelable** - We should be able to cancel any billing gracefully
+- **Terminate-able** - We should be able to rip the carpet out from a billing routine
+- **Nonblocking failures** - A failure of one billing workflow should not halt the execution of others.
+
```

Comment:
Although it's obvious for both of us, I'd explicitly add an **Observable** requirement: we want to alert on tasks being delayed too long / dropped, and we want to identify the cause quickly and reliably.